### PR TITLE
Add CEL Parser

### PR DESCRIPTION
**Pull request by matthchr**
* Includes json_type_provider for translating between v1.JSON and map[string]any in CEL.
* Includes cache for CEL expressions and CEL Envs (one env per resource type).

Related to #2555 

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains tests
- [ ] this PR contains YAML Samples

### CODE REVIEW

File: v2/internal/util/cel/env_cache.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	asometrics "github.com/Azure/azure-service-operator/v2/internal/metrics"
+)
+
+type envCacheItem struct {
+	env *cel.Env
+}
+
+type EnvCache struct {
```
Comment (matthchr) [inline v2/internal/util/cel/env_cache.go:25]: Missing some comments

File: v2/internal/util/cel/compilation.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/google/cel-go/cel"
+	"github.com/pkg/errors"
+)
+
+var (
+	StringType = cel.StringType
+	MapType    = cel.MapType(cel.StringType, cel.StringType)
+)
+
+func AllowedOutputTypes() []*cel.Type {
```
Comment (matthchr) [inline v2/internal/util/cel/compilation.go:18]: Comments?

File: v2/internal/util/cel/cel.go [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+
+// Note: if we ever make a breaking change here, we should allow users to control which "version" of CEL parsing they
+// Note: get with an annotation or property on the operatorSpec.
+
+const (
+	SelfIdent   = "self"
+	SecretIdent = "secret"
+)
+
+//
```
Comment (matthchr) [inline v2/internal/util/cel/cel.go:32]: Fix

File: v2/internal/util/cel/cel.go [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+//	ExpressionClassificationMap = ExpressionClassification("map")
+//)
+
+var evaluator ExpressionEvaluator
+
+func Evaluator() ExpressionEvaluator {
+	return evaluator
+}
+
+// TODO: We could put this instead in genruntime if we thought it made more sense to keep the "static value for use by webhooks"
```
Comment (matthchr) [inline v2/internal/util/cel/cel.go:46]: Curious as to folks thoughts on this...

File: v2/internal/util/cel/cel.go [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+	if result.programCache == nil {
+		// Return an error if metrics wasn't configured
+		envCache := NewEnvCache(result.metrics, result.log, NewEnv)
+		result.programCache = NewProgramCache(envCache, result.metrics, result.log, Compile)
+	}
+
+	return result, nil
+}
+
+// TODO: Need to use this someplace
```
Comment (matthchr) [inline v2/internal/util/cel/cel.go:126]: Remove, it's used now.

File: v2/internal/util/cel/cel.go [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+		SelfIdent:   self,
+		SecretIdent: secret,
+	}
+	// TODO: We may want to use ContextEval here, alongside prgm, err := env.cel.Program(ast, cel.InterruptCheckFrequency(10))
+	out, _, err := program.Program.Eval(input)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to eval CEL expression: %q", expression)
+	}
+
+	// TODO: When inputs were Dyn, it was hard to statically typecheck based on out.Type. We technically can now
```
Comment (matthchr) [inline v2/internal/util/cel/cel.go:214]: This isn't really a TODO

File: v2/internal/util/cel/cel.go (hunk 5 above)
Comment (matthchr) [reply v2/internal/util/cel/cel.go:126]: or will be* in a subsequent PR.

File: v2/internal/util/cel/cel_test.go [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+					r.Spec.Break = " there"
+					r.Spec.Loop = " friend"
+				}),
+			expression:  `self.spec.const + self.spec.break + self.spec.loop`,
+			expectedStr: "hello there friend",
+		},
+	}
+
+	for _, c := range cases {
+		c := c
```
Comment (theunrepentantgeek) [inline v2/internal/util/cel/cel_test.go:314]: YIL (yesterday-I-learned) that this is no longer needed; as of [Go 1.22](https://tip.golang.org/doc/go1.22#language), loop variables are created new for each iteration.

File: v2/internal/util/cel/json_type_provider.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+const (
+	typeName = "v1.JSON"
+)
+
+var jsonMapType = types.NewMapType(types.StringType, types.NewObjectType(typeName))
+
+// TODO: Tests
```
Comment (theunrepentantgeek) [inline v2/internal/util/cel/json_type_provider.go:24]: Still pending?

File: v2/internal/util/cel/json_type_provider.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+	_ types.Provider = &jsonProvider{}
+)
+
+// types.Adapter impl
+func (j *jsonProvider) NativeToValue(value any) ref.Val {
+	if value == nil {
+		return types.NullValue
+	}
+
+	// TODO: NO idea
```
Comment (theunrepentantgeek) [inline v2/internal/util/cel/json_type_provider.go:75]: :grin:

File: v2/internal/util/cel/program_cache.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+func (c *ProgramCache) Get(resource reflect.Type, expression string) (*CompilationResult, error) {
+	envKey := getTypeImportPath(resource)
+
+	env, err := c.envCache.Get(resource)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get CEL env")
+	}
+	key := fmt.Sprintf("%s-%s", envKey, expression)
+
+	// TODO: Do we also want to cache errors?
```
Comment (theunrepentantgeek) [inline v2/internal/util/cel/program_cache.go:92]: Probably, yes.

**Review by theunrepentantgeek — APPROVED** [review]

File: v2/internal/util/cel/cel.go [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+	Check(expression string, self any) (*cel.Type, error)
+	FindSecretUsage(expression string, self any) (set.Set[string], error)
+	Start()
+	Stop()
+}
+
+type ExpressionEvaluatorOption func(e *expressionEvaluator) (*expressionEvaluator, error)
+
+// Cache configures the program cache for the expression evaluator.
+// The program cache is used to avoid re-evaluating expressions or Envs
```
[the remark this answers is not in the document]
Comment (matthchr) [reply v2/internal/util/cel/cel.go:59]: compiling, fixed.

File: v2/internal/util/cel/cel.go [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+			name := splits[0]
+			return name
+		}
+	}
+
+	return field.Name
+}
+
+func escapeFieldName(field string) string {
+	replacer := strings.NewReplacer("__", "__underscores__", ".", "__dot__", "-", "__dash__", "/", "__slash__")
```
[the remark this answers is not in the document]
Comment (matthchr) [reply v2/internal/util/cel/cel.go:272]: Great suggestion.

File: v2/internal/util/cel/cel_test.go (hunk 7 above)
Comment (matthchr) [reply v2/internal/util/cel/cel_test.go:314]: I actually "knew" this but hadn't really registered that it meant I could avoid it in ASO.

Removed!

File: v2/internal/util/cel/json_type_provider.go (hunk 8 above)
Comment (matthchr) [reply v2/internal/util/cel/json_type_provider.go:24]: Not really, it's covered by the tests we have now. This was from when I originally wrote this and didn't have all the tests yet.

File: v2/internal/util/cel/program_cache.go (hunk 10 above)
Comment (matthchr) [reply v2/internal/util/cel/program_cache.go:92]: I've updated to cache errors here, and updated the test to ensure it works.

**Pull request merged**